calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -140,33 +140,46 @@ func removeHostTunnelAddr(ctx context.Context, c client.Interface, nodename stri
 	}
 }
 
+// Make best effort to release IPAM assigned IP address.
+// Log, but otherwise ignore errors.
+func releaseIPAddrByHandle(ctx context.Context, c client.Interface, handle string, ipv4Addr string) {
+	err := c.IPAM().ReleaseByHandle(ctx, handle)
+	if err != nil {
+		log.WithError(err).WithField("IP", ipv4Addr).Error("Error releasing IP address on failure")
+	} else {
+		log.WithField("IP", ipv4Addr).Debug("Released IP address")
+	}
+}
+
 // assignHostTunnelAddr claims an IPIP-enabled IP address from the first pool
 // with some space. Stores the result in the host's config as its tunnel
 // address.
 func assignHostTunnelAddr(ctx context.Context, c client.Interface, nodename string, ipipCidrs []net.IPNet) {
+	handle := "ipipAddr:" + nodename
```

Comment:
So the case that I was worried about was:
-  calico/node starts, knows it needs an IPIP Addr
-  IPIP address is assigned
-  calico/node restarts before the IP address is set on the node resource

What that means is on the next restart, that IP address will be "lost".  

So, to find that "lost" IP address, I think before we assign a new address, we should first check to see if there is an IP address assigned to the handle, and if so check if the IP is still valid (based on the IPIP pools) and if not release it - otherwise use the one that was previously assigned.

Reply:
Kinda feel that deleting a node should also delete the IPIP address by handle (rather than relying on the IPIP Tunnel Address field being set correctly) -- but that distributes where this code lives which doesn't feel correct.